netket/experimental/operator/_particle_number_conserving_fermionic/_kernels.py

```diff
@@ -0,0 +1,389 @@
+# this file contains the logic of the operators, essentially _jw_kernel
+
+from typing import Union
+
+from functools import partial
+
+import numpy as np
+import itertools
+
+import jax
+import jax.numpy as jnp
+
+from netket.jax import reduce_xor, COOArray
+from netket.utils.types import Array
+
+from ._operator_data import PNCOperatorDataType
+
+
+def _comb(kl: Array, n: int) -> Array:
+    """
+    compute all combinations of n elements from a set kl
+    Args:
+        kl: 1d array of elements
+        n: size n
+    Returns:
+        an array containg all combinations of size n of elemenets in kl
+    """
+    if len(kl) < n:
+        return jnp.zeros((n, 0), dtype=kl.dtype)
+    c = list(itertools.combinations(np.arange(len(kl)), n))
+    return kl[np.array(c, dtype=kl.dtype).T[::-1]]
+
+
+def _jw_kernel(
+    k_destroy: Array, l_create: Array, x: Array
+) -> tuple[Array, Array, Array]:
+    """
+    compute all matrix elements :math:`x^\\prime` such that :math:`\\langle x^\\prime | \\hat c^\\dagger_{l_m} \\cdots \\hat c^\\dagger_{l_1} \\hat c_{k_n} \\cdots \\hat c_{k_1} | x\rangle \neq 0`
+    of a batch of states x.
+
+    Args:
+        k_destroy: an array of indices of the :math:`\\hat c`
+        l_create: an array of indices of the :math:`\\hat c^\\dagger`
+        x: an array of states
+    Returns:
+        xp: matrix elements :math:`x^\\prime`
+        sign: sign (value) of the matrix element if it is nonzero, arbitrary value otherwise
+        create_was_empty: if the matrix element is zero
+    """
+    # destroy
+    xd = jax.vmap(lambda i: x.at[i].set(0))(k_destroy.T)
+    # create
+    xp = jax.vmap(jax.vmap(lambda x, i: x.at[i].set(1), in_axes=(None, 0)))(
+        xd, l_create
+    )
+
+    m = jnp.arange(x.shape[-1], dtype=k_destroy.dtype)
+
+    # we apply the destruction operators in descending order,
+    # the jordan-wigner sign of an operator does not depend on sites larger than it, therefore,
+    # given it is in normal order, we can compute it all in terms of the initial state.
+    # (sum the axis which is the one of the indices we destroy/create (size number of operators//2))
+    jw_mask_destroy = reduce_xor(k_destroy[..., None] > m, axes=0)
+
+    # same for when we create again, except then have to apply it to the state where we already destroyed
+    jw_mask_create = reduce_xor(l_create[..., None] > m, axes=2)
+
+    create_was_empty = jax.vmap(jax.vmap(lambda x, i: ~x[i].any(), in_axes=(None, 0)))(
+        xd, l_create
+    )
+
+    sgn_destroy = reduce_xor(jw_mask_destroy * x[None], axes=-1)
```

Comment:
I'm not a fan of relying on jax internals which might stop working or be changed in the future if we can avoid it...
Does this not work?
I'll admit this is slightly less readable but at least we don't have to worry about jax changing things in the future.
```suggestion
    sgn_destroy = jax.lax.reduce_xor(jw_mask_destroy * x[None], axes=x.ndim)
```